controller/backing_image_manager_controller.go

```diff
@@ -0,0 +1,1233 @@
+	defer logrus.Infof("Shutting down Longhorn backing image manager controller")
+
+	if !cache.WaitForNamedCacheSync("longhorn backing image manager", stopCh, c.bimStoreSynced, c.biStoreSynced, c.nStoreSynced, c.pStoreSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *BackingImageManagerController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *BackingImageManagerController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncBackingImageManager(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *BackingImageManagerController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if c.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("Error syncing Longhorn backing image manager %v: %v", key, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logrus.Warnf("Dropping Longhorn backing image manager %v out of the queue: %v", key, err)
+	c.queue.Forget(key)
+}
+
+func getLoggerForBackingImageManager(logger logrus.FieldLogger, bim *longhorn.BackingImageManager) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"backingImageManager": bim.Name,
+			"nodeID":              bim.Spec.NodeID,
+			"diskUUID":            bim.Spec.DiskUUID,
+		},
+	)
+}
+
+func (c *BackingImageManagerController) syncBackingImageManager(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "BackingImageManagerController failed to sync %v", key)
+	}()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != c.namespace {
+		return nil
+	}
+
+	bim, err := c.ds.GetBackingImageManager(name)
+	if err != nil {
+		if !datastore.ErrorIsNotFound(err) {
+			c.logger.WithField("backingImageManager", name).WithError(err).Error("Failed to retrieve backing image manager from datastore")
+			return err
+		}
+		c.logger.WithField("backingImageManager", name).Debug("Can't find backing image manager, may have been deleted")
+		return nil
+	}
+
+	log := getLoggerForBackingImageManager(c.logger, bim)
+
+	if !c.isResponsibleFor(bim) {
+		return nil
+	}
+	if bim.Status.OwnerID != c.controllerID {
+		bim.Status.OwnerID = c.controllerID
+		bim, err = c.ds.UpdateBackingImageManagerStatus(bim)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Debugf("BackingImageManagerController on node %v picked up backing image manager", c.controllerID)
+	}
+
+	if bim.DeletionTimestamp != nil {
+		if err := c.cleanupBackingImageManager(bim); err != nil {
+			return err
+		}
+		pod, err := c.ds.GetPod(bim.Name)
+		if err != nil {
+			return err
+		}
+		if pod != nil {
+			return nil
+		}
```

Comment:
Does the controller need to wait for the pod cleanup before removing the finalizer? If YES, I may need to add force deletion in case of node down.